### PR TITLE
Addition of histogram scattering treatment to openmc.mgxs.Library 

### PR DESCRIPTION
**Pull request by nelsonag**
OpenMC's multi-group mode supports scattering kernels represented as a Legendre expansion, a tabular f(mu), and histogram bins of f(mu).  This PR adds functionality to openmc.mgxs.Library et al to aid in automagically generating the histogram bin data from OpenMC's continuous-energy mode for MG mode.  Incorporation of this PR will check two more boxes off in Issue #597.

Two other minor changes included in this PR are:
- Simlpifications to the MuFilter, AzimuthalFilter, PolarFilter classes in the PyAPI to inherit from RealFilter just like EnergyFilter does.
- Corrected the scatter_format from `[Order][G][G']` to `[G][G'][Order]` (Ughhh this row-major vs column-major translation has obviously been killing me)
- Updated the openmc-update-mgxs script to incorporate changes from recent PRs (i.e., the use of xs_shapes and converting to eV).

### CODE REVIEW

File: openmc/mgxs/mgxs.py [hunk 1]
```diff
@@ -3755,28 +3845,38 @@ def get_pandas_dataframe(self, groups='all', nuclides='all', moment='all',
...
+            if moment != 'all':
+                cv.check_type('moment', moment, Integral)
+                cv.check_greater_than('moment', moment, 0, equality=True)
+                cv.check_less_than(
+                    'moment', moment, self.legendre_order, equality=True)
+                df = df[df['moment'] == 'P{}'.format(moment)]
+
+        elif self.scatter_format == 'histogram':
+            # Add a change-in-angle (mu) column to dataframe
+            ###TODO NOT SURE I NEED TO DO THIS
```
Comment (paulromano) [inline openmc/mgxs/mgxs.py]: I would say yes.

File: examples/python/pincell_multigroup/build-xml.py [hunk 2]
```diff
@@ -26,14 +28,16 @@
...
-uo2_xsdata.set_scatter_matrix(
+scatter_matrix = np.array(
     [[[0.1275370, 0.0423780, 0.0000094, 0.0000000, 0.0000000, 0.0000000, 0.0000000],
       [0.0000000, 0.3244560, 0.0016314, 0.0000000, 0.0000000, 0.0000000, 0.0000000],
       [0.0000000, 0.0000000, 0.4509400, 0.0026792, 0.0000000, 0.0000000, 0.0000000],
       [0.0000000, 0.0000000, 0.0000000, 0.4525650, 0.0055664, 0.0000000, 0.0000000],
       [0.0000000, 0.0000000, 0.0000000, 0.0001253, 0.2714010, 0.0102550, 0.0000000],
       [0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.0012968, 0.2658020, 0.0168090],
       [0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.0085458, 0.2730800]]])
+scatter_matrix = np.swapaxes(np.swapaxes(scatter_matrix, 0, 1), 1, 2)
```
Comment (nelsonag) [inline examples/python/pincell_multigroup/build-xml.py]: - [x]

File: openmc/filter.py [hunk 3]
```diff
@@ -102,7 +102,8 @@ def _recursive_subclasses(cls):
         """Return all subclasses and their subclasses, etc."""
         subs = cls.__subclasses__()
         subsubs = [grand for s in subs for grand in s.__subclasses__()]
-        return subs + subsubs
+        subsubsubs = [grand for s in subsubs for grand in s.__subclasses__()]
+        return subs + subsubs + subsubsubs
```
Comment (nelsonag) [inline openmc/filter.py]: - [x]

File: openmc/mgxs/mgxs.py [hunk 4]
```diff
@@ -3700,11 +3777,24 @@ def get_xs(self, in_groups='all', out_groups='all',
...
+            # so 1-group problems have the correct shape.
+            if self.scatter_format == 'histogram':
+                axes = (5, 4, 0)
+            else:
+                axes = (4, 3, 0)
+            # Squeeze will return a ValueError if the axis has a size greater
+            # than 1, so try each axis in axes one at a time, catching the
+            # ValueError as needed.
+            for axis in axes:
+                try:
```
Comment (nelsonag) [inline openmc/mgxs/mgxs.py]: - [x]

**Pull request merged**